From 4b66288d9c102fceee3295996f32b0eaad21687b Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 8 Apr 2024 17:45:23 +0200
Subject: [PATCH 0812/1695] drivers: stm32_tamp: handle LSE monitoring tamper

When a LSE monitoring tamper occurs, there is a specific sequence to
follow to disable LSE CSS and LSE clock. After that, a backup domain
reset is mandatory as well. Any other specific handling is implementation
specific.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1bb33048a5acfd9fe618a0647ea3d732040d1619 drivers: stm32_tamp: handle LSE monitoring tamper]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I44cd95ad99a36a1866a40d18750128bd298d4982
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/372209
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/stm32_tamp.c | 42 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/core/drivers/stm32_tamp.c b/core/drivers/stm32_tamp.c
index 0bfc7b72e..3577972a5 100644
--- a/core/drivers/stm32_tamp.c
+++ b/core/drivers/stm32_tamp.c
@@ -13,6 +13,18 @@
 #include <drivers/stm32_gpio.h>
 #include <drivers/stm32_rtc.h>
 #include <drivers/stm32_tamp.h>
+#if defined(CFG_STM32MP15)
+#include <drivers/stm32mp1_rcc.h>
+#endif
+#if defined(CFG_STM32MP13)
+#include <drivers/stm32mp13_rcc.h>
+#endif
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23)
+#include <drivers/stm32mp25_rcc.h>
+#endif
+#if defined(CFG_STM32MP21)
+#include <drivers/stm32mp21_rcc.h>
+#endif
 #include <io.h>
 #include <kernel/boot.h>
 #include <kernel/delay.h>
@@ -1414,6 +1426,32 @@ static void __maybe_unused stm32_tamp_lock_boot_hardware_key(void)
 		io_setbits32(base + _TAMP_SECCFGR, _TAMP_SECCFGR_BHKLOCK);
 }
 
+static void stm32_tamp_handle_lse_tamper(void)
+{
+	vaddr_t rcc_base = stm32_rcc_base();
+
+	/* Disable the LSE CSS feature */
+	io_clrbits32(rcc_base + RCC_BDCR, RCC_BDCR_LSECSSON);
+
+	/* Disable the LSE clock */
+	io_clrbits32(rcc_base + RCC_BDCR, RCC_BDCR_LSEON);
+
+	/* Clear the associated error flag */
+#if defined(CFG_STM32MP13) || defined(CFG_STM32MP15)
+	io_setbits32(rcc_base + RCC_MP_CIFR, RCC_MP_CIFR_LSECSSF);
+#else
+	if (io_read32(rcc_base + RCC_C1CIFCLRR) & RCC_C1CIFCLRR_LSECSSF)
+		io_setbits32((rcc_base + RCC_C1CIFCLRR), RCC_C1CIFCLRR_LSECSSF);
+#endif
+
+	/*
+	 * Fixme: Add logic to handle the LSE tamper here (e.g change RTC clock
+	 * source instead). This part is implementation specific.
+	 */
+
+	io_clrbits32(rcc_base + RCC_BDCR, RCC_BDCR_RTCCKEN);
+}
+
 static enum itr_return stm32_tamp_it_handler(struct itr_handler *h __unused)
 {
 	vaddr_t base = io_pa_or_va(&stm32_tamp.pdata.base, 1);
@@ -1451,6 +1489,10 @@ static enum itr_return stm32_tamp_it_handler(struct itr_handler *h __unused)
 				io_setbits32(base + _TAMP_SCR,
 					     _TAMP_SCR_ITAMP(id));
 
+			/* Handler LSE monitoring tamper */
+			if (id - INT_TAMP1 + 1 == INT_TAMPER_LSE_MONITORING)
+				stm32_tamp_handle_lse_tamper();
+
 			if (ret & TAMP_CB_RESET)
 				do_reset("TAMPER");
 		}
-- 
2.47.2

